### DISCUSSION TITLE
proxy route rule issue with firebase hosting

### DISCUSSION THREAD

**Discussion by davidstackio** [nuxt/nuxt#25014] [Questions] [answered]
I'm trying to proxy [PostHog as instructed here](https://posthog.com/docs/advanced/proxy/nuxt) and I'm getting this 500 error in my Google Cloud Functions logs:

```
[nuxt] [request error] [unhandled] [500] fetch failed
at Object.fetch (node:internal/deps/undici/undici:11372:11)  
at async sendProxy (./chunks/nitro/firebase-gen-2.mjs:1998:20)  
at async Object.handler (./chunks/nitro/firebase-gen-2.mjs:2354:19)  
at async toNodeHandle (./chunks/nitro/firebase-gen-2.mjs:2543:7)
```

GCP function logs:
![image](https://github.com/nuxt/nuxt/assets/2013388/6d0198b3-2528-4229-97df-083cd8845539)

Corresponding browser console logs:
![image](https://github.com/nuxt/nuxt/assets/2013388/f6049424-5971-46eb-8149-ca9480988ebd)

@pi0 suggested I post a discussion here (https://github.com/unjs/nitro/issues/1717) and said it may be related to the headers.

Here are the headers from one of the requests:
![image](https://github.com/nuxt/nuxt/assets/2013388/0c285cda-2159-4b0a-9b95-67d885b92bb0)

And the payload (I don't think I needed to block this out since it looks encrypted, but I'm still learning :)):
![image](https://github.com/nuxt/nuxt/assets/2013388/f1564e87-e208-4dd6-a8c2-6c0c13757bf8)

**Comment by pi0** [comment]
Thanks for moving it here. Would you be able to prepare a runnable reproduction or somehow share your or an example project setup? This way I can quickly try and figure out if anything might be wrong in either configuration or Nuxt/Nitro codebase. (in chance I missed this thread don't hesitate to ping me @ pi0 on dicord!

**Reply by davidstackio** [reply]
@pi0 I just messaged you on Discord, but here's my Nuxt config file:

```ts
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  extends: ["../nuxt-base"],
  nitro: {
    preset: "firebase",
    firebase: {
      gen: 2,
      nodeVersion: "18",
    },
  },

  // Modules
  modules: ["@nuxt/content", "@pinia/nuxt", "@nuxtseo/module"],

  // Module config
  content: {
    markdown: {
      anchorLinks: false,
    },
  },

  // SEO
  site: {
    url: "https://[mydomain].com",
    name: "Name",
    description: "description",
    defaultLocale: "en",
    identity: {
      type: "Organization",
    },
  },
  sitemap: {
    strictNuxtContentPaths: true,
  },
  ogImage: {
    enabled: false,
  },

  // App config
  app: {
    head: {
      titleTemplate: "%pageTitle",
    },
  },

  // Runtime config
  runtimeConfig: {
    // Config within public will be also exposed to the client
    public: {
      posthog: {
        publicKey: "[mykey]",
        host: "https://[mydomain].com/ingest",
      },
    },
  },
});
```

And the posthog plugin I'm using (NOTE: this is in the nuxt-base layer):
```ts
import { posthog } from "posthog-js";

export default defineNuxtPlugin(() => {
  if (process.env.NODE_ENV === "production") {
    const config = useRuntimeConfig();

    // Initialize PostHog
    posthog.init(config.public.posthog.publicKey, {
      api_host: config.public.posthog.host || "https://app.posthog.com",
      ui_host: config.public.posthog.host
        ? "https://app.posthog.com"
        : undefined,
      loaded: (posthog) => {
        if (process.env.NODE_ENV === "development") posthog.debug();
      },
      session_recording: {
        maskAllInputs: false,
      },
    });

    // Make sure that pageviews are captured with each route change
    if (!posthog.has_opted_out_capturing()) {
      const router = useRouter();
      router.afterEach((to) => {
        posthog.capture("$pageview", {
          current_url: to.fullPath,
        });
      });
    }
  }
});
```

And the nuxt config for the nuxt-base layer:
```ts
// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  devtools: { enabled: true },
  typescript: {
    shim: false,
  },

  // Modules
  modules: ["vuetify-nuxt-module"],

  // Runtime config
  runtimeConfig: {
    // Config within public will be also exposed to the client
    public: {
      posthog: {
        publicKey: "",
        host: "",
      },
    },
  },

  // Routes
  routeRules: {
    // Posthog reverse proxy
    "/ingest/**": { proxy: "https://app.posthog.com/**" },
  },
});

```

FYI Posthog works fine until I try to do the proxy.

**Reply by TheAlexLichter** [reply]
Could you just use a custom nitro endpoint + `proxyRequest` as e.g. [shown here](https://www.youtube.com/watch?v=J4E5uYz5AY8)? 🤔

**Reply by pi0** [reply]
This and most of similar proxy-domain cases should out of the box with a simple `proxy` rule. Currently investigating why (have a reproduction now 👍🏼 )

**Reply by davidstackio** [reply]
@manniL Thanks for the video. Checked out some other ones too and just subscribed :) My use case doesn't need the host changed at runtime (I just did that in my code examples for privacy) and since the event is sending from a 3rd party library (posthog), I'm not sure if I'd even be able to manually send the body of a POST request using `proxyRequest` (or maybe I'm missing something).

Either way, I'd expect the routeRules with a proxied POST request to work out of the box. At the moment it looks like there may be a bug with that, which it looks like @pi0 is investigating. Happy to provide any additional information to help with this process!

**Reply by TheAlexLichter** [reply]
Also POST request works fine for me with the following setup (POSTing to the jsonplaceholder API):

`server/api/test.ts`

```ts
export default defineEventHandler((event) => {
  return proxyRequest(event, 'https://jsonplaceholder.typicode.com/posts')
});
```

`app.vue`

```vue
<script setup lang="ts">
async function send () {
  const result = await $fetch('/api/test', {
    method: 'POST',
    body: {
      title: 'foo',
      body: 'bar',
      userId: 1,
    },
    headers: {
      'Content-type': 'application/json; charset=UTF-8',
    },
  })
  console.log(result)
}
</script>

<template>
  <div>
    Hey hey hey!
    <button @click="send">Send</button>
  </div>
</template>
```

<img width="1300" alt="image" src="https://github.com/nuxt/nuxt/assets/640208/1e7660e0-0d59-404f-82e8-b77930fa377b">

**Reply by pi0** [reply]
(off topic) @manniL issue was with firebase hosting (https://github.com/unjs/h3/pull/604). `proxy` route rule acts exactly the same as above and under the hood also exactly uses same `proxyRequest>sendRequest` utils and (unless really required) I would always recommend route rules over programmatic API as it can guarantee more stability.

**Comment by pi0** [comment, marked as answer]
Issue were with h3 wrongly encoding already parsed body from firebase runtime. https://github.com/unjs/h3/pull/604

Simple workaround until h3 releases: (tested against firebase emulator waiting for confirming from David as well

server/plugins/firebase.ts: 
```ts
export default defineNitroPlugin((nitro) => {
  nitro.hooks.hook('request', (event) => {
    // @ts-ignore https://github.com/unjs/h3/pull/604
    event.node.req.body = event.node.req.rawBody
  })
})
```

**Reply by davidstackio** [reply]
Adding this fixed the issue. Thank you!
